Freertos initial

rbudai98 opened the pull request:
* FreeRTOS generic makefile added
* Current build commands need: `FREERTOS=y` for enabling the library and `FREERTOS_KERNEL=/path/to/Kernel` for specifying the path towards freeRTOS Kernel (https://github.com/FreeRTOS/FreeRTOS-Kernel)
* FreeRTOS.config file is linked from project folder
* Current implementation has the ARM4F CMSIS file linked for max devices, in future it will need generalizations
* (example project in different PR)

rbudai98 commented:
v2:
* added mutex includes to projects 
* corrected doxygen for platform/freertos files

buha commented on include/no_os_eeprom.h:
```diff
@@ -79,6 +81,8 @@ struct no_os_eeprom_desc {
 	/** EEPROM extra parameters (device specific) */
 	void *extra;
+	/** EEPROM mutex*/
+	void 		*mutex;
```
I would put the mutex as the first member of this structure to be able to lock the resource by a simple cast. It might be useful in the future.

buha commented on drivers/platform/freeRTOS/freertos_semaphore.c:
```diff
@@ -0,0 +1,84 @@
+__attribute__((weak)) inline void no_os_semaphore_take(void *semaphore)
+{
+	xSemaphoreTake(semaphore,
+		       freeRTOS_WAIT_TIME); //replace with define from freertos config file
```
you can remove the comment I think

buha commented on tools/scripts/freertos.mk:
```diff
@@ -0,0 +1,34 @@
+        $(FREERTOS_KERNEL)/queue.c \
+        $(FREERTOS_KERNEL)/timers.c 
+
+RTOS_SRCS += $(FREERTOS_KERNEL)/portable/GCC/ARM_CM4F/port.c
```
At some point in the future it would be nice if this was selected automatically based on processor because as it is now, it will certainly cause troubles when attempting a build with FreeRTOS for CM0 or CM33 etc.

Unfortunately there is no trivial way to do this, just by string manipulation/regex.

rbudai98 replied:
As per discussed it will be solved in a different PR

CiprianRegus commented on drivers/api/no_os_eeprom.c:
```diff
@@ -121,5 +127,9 @@ int32_t no_os_eeprom_read(struct no_os_eeprom_desc *desc, uint32_t address,
+
 	return desc->platform_ops->read(desc, address, data, bytes);
+
+	no_os_mutex_unlock(desc->mutex);
```
this will cause the mutex to remain locked. Return after unlocking.

CiprianRegus commented on drivers/api/no_os_i2c.c:
```diff
@@ -105,8 +107,12 @@ int32_t no_os_i2c_write(struct no_os_i2c_desc *desc,
 
-	return desc->platform_ops->i2c_ops_write(desc, data, bytes_number,
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->i2c_ops_write(desc, data, bytes_number,
```
please declare `ret` at the start of the function. There are some other places where this applies.

CiprianRegus commented on drivers/api/no_os_spi.c:
```diff
@@ -125,14 +132,21 @@ int32_t no_os_spi_transfer(struct no_os_spi_desc *desc,
 
+	no_os_mutex_lock(desc->mutex);
+
 	for (i = 0; i < len; i++) {
-		if (msgs[i].rx_buff != msgs[i].tx_buff || !msgs[i].tx_buff)
+		if (msgs[i].rx_buff != msgs[i].tx_buff || !msgs[i].tx_buff) {
+			no_os_mutex_unlock(desc->mutex);
 			return -EINVAL;
+		}
 		ret = no_os_spi_write_and_read(desc, msgs[i].rx_buff,
 					       msgs[i].bytes_number);
-		if (NO_OS_IS_ERR_VALUE(ret))
+		if (NO_OS_IS_ERR_VALUE(ret)) {
+			no_os_mutex_unlock(desc->mutex);
 			return ret;
+		}
 	}
 
+	no_os_mutex_unlock(desc->mutex);
 	return 0;
```
this may be simplified by using `goto` statements as such:

```
for(...) {
ret = no_os_spi_write_and_read(desc, msgs[i].rx_buff,
					       msgs[i].bytes_number);
if (NO_OS_IS_ERR_VALUE(ret))
        goto out;
}

out:
        no_os_mutex_unlock(desc->mutex);

return ret;
```

CiprianRegus commented on include/no_os_mdio.h:
```diff
@@ -74,6 +74,8 @@ struct no_os_mdio_init_param {
 	/** Platform or implementation specific parameters. */
 	void *extra;
+	/** MDIO mutex*/
+	void 		*mutex;
```
I think there may be the case where the user declares more than 1 descriptor for a peripheral. By placing the mutex in here, it won't be shared between descriptors, which might cause issues if 2 descriptors are used in 2 different threads. One possible solution would be to declare it as `static` in the corresponding source files (`no_os_mdio.c` in this case).

rbudai98 replied:
In general it is correct. However if we switch to `static` declaration it won't be possible to declare for instance two different SPI descriptors, that might have different ports and might be used simultaneously.

rbudai98 replied:
Modified i2c and spi api to support bus descriptor, will add there

amiclaus commented on drivers/api/no_os_i2c.c:
```diff
@@ -130,6 +136,10 @@ int32_t no_os_i2c_read(struct no_os_i2c_desc *desc,
-	return desc->platform_ops->i2c_ops_read(desc, data, bytes_number,
-						stop_bit);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->i2c_ops_read(desc, data, bytes_number,
```
variable declaration at the start of the function.

amiclaus commented on drivers/api/no_os_mdio.c:
```diff
@@ -101,7 +104,11 @@ int no_os_mdio_write(struct no_os_mdio_desc *desc, uint32_t reg, uint16_t val)
 
-	return desc->ops->write(desc, reg, val);
+	no_os_mutex_lock(desc->mutex);
+	int ret = desc->ops->write(desc, reg, val);
```
ditto.

amiclaus commented on drivers/api/no_os_mdio.c:
```diff
@@ -121,5 +128,9 @@ int no_os_mdio_read(struct no_os_mdio_desc *desc, uint32_t reg, uint16_t *val)
 
-	return desc->ops->read(desc, reg, val);
+	no_os_mutex_lock(desc->mutex);
+	int ret = desc->ops->read(desc, reg, val);
```
ditto

amiclaus commented on drivers/api/no_os_pwm.c:
```diff
@@ -97,7 +100,9 @@ int32_t no_os_pwm_enable(struct no_os_pwm_desc *desc)
 
+	no_os_mutex_lock(desc->mutex);
 	return desc->platform_ops->pwm_ops_enable(desc);
+
```
unrelated change.

amiclaus commented on drivers/api/no_os_spi.c:
```diff
@@ -41,6 +41,8 @@
 #include <stdlib.h>
 #include "no_os_error.h"
+#include "no_os_mutex.h"
+
```
unrelated change.

amiclaus commented on drivers/api/no_os_spi.c:
```diff
@@ -102,7 +105,11 @@ int32_t no_os_spi_write_and_read(struct no_os_spi_desc *desc,
 
-	return desc->platform_ops->write_and_read(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret =  desc->platform_ops->write_and_read(desc, data, bytes_number);
```
ditto.

amiclaus commented on drivers/api/no_os_spi.c:
```diff
@@ -125,14 +132,21 @@ int32_t no_os_spi_transfer(struct no_os_spi_desc *desc,
 	for (i = 0; i < len; i++) {
-		if (msgs[i].rx_buff != msgs[i].tx_buff || !msgs[i].tx_buff)
+		if (msgs[i].rx_buff != msgs[i].tx_buff || !msgs[i].tx_buff) {
+			no_os_mutex_unlock(desc->mutex);
```
this could be done smoother with a conditional jump.

amiclaus commented on drivers/api/no_os_uart.c:
```diff
@@ -117,7 +119,11 @@ int32_t no_os_uart_read(struct no_os_uart_desc *desc,
 
-	return desc->platform_ops->read(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->read(desc, data, bytes_number);
```
ditto

amiclaus commented on drivers/api/no_os_uart.c:
```diff
@@ -137,7 +143,11 @@ int32_t no_os_uart_write(struct no_os_uart_desc *desc,
 
-	return desc->platform_ops->write(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->write(desc, data, bytes_number);
```
ditto.

amiclaus commented on drivers/api/no_os_uart.c:
```diff
@@ -157,7 +167,11 @@ int32_t no_os_uart_read_nonblocking(struct no_os_uart_desc *desc,
 
-	return desc->platform_ops->read_nonblocking(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->read_nonblocking(desc, data, bytes_number);
```
ditto

amiclaus commented on drivers/api/no_os_uart.c:
```diff
@@ -177,7 +191,11 @@ int32_t no_os_uart_write_nonblocking(struct no_os_uart_desc *desc,
 
-	return desc->platform_ops->write_nonblocking(desc, data, bytes_number);
+	no_os_mutex_lock(desc->mutex);
+	int32_t ret = desc->platform_ops->write_nonblocking(desc, data, bytes_number);
```
ditto

amiclaus commented on drivers/platform/freeRTOS/freertos_delay.c:
```diff
@@ -0,0 +1,85 @@
+
+void no_os_udelay(uint32_t usecs)
+{
+	vTaskDelay(((usecs/1000000) * configTICK_RATE_HZ));
```
nit: spaces before and after `/`

amiclaus commented on drivers/platform/freeRTOS/freertos_delay.c:
```diff
@@ -0,0 +1,85 @@
+ */
+void no_os_mdelay(uint32_t msecs)
+{
+	vTaskDelay(((msecs/1000) * configTICK_RATE_HZ));
```
ditto: spaces before and after `/`

amiclaus commented on drivers/platform/freeRTOS/freertos_delay.c:
```diff
@@ -0,0 +1,85 @@
+struct no_os_time no_os_get_time(void)
+{
+	unsigned long long Xtime_Global;
+	float fractional_part = 0;
```
can we avoid `float`?

amiclaus commented on drivers/platform/freeRTOS/freertos_semaphore.c:
```diff
@@ -0,0 +1,84 @@
+{
+	no_os_free(semaphore);
+}
+
```
nit: remove line.

amiclaus commented on projects/ad738x_fmcz/src.mk:
```diff
@@ -32,4 +33,6 @@ INCS +=	$(INCLUDE)/no_os_axi_io.h \
-	$(INCLUDE)/no_os_alloc.h
+	$(INCLUDE)/no_os_alloc.h \
+	$(INCLUDE)/no_os_mutex.h 
+
```
nit: unrelated change.

amiclaus commented on projects/adxrs290-pmdz/src.mk:
```diff
@@ -27,14 +27,16 @@ INCS += $(INCLUDE)/no_os_delay.h        \
         $(NO-OS)/util/no_os_lf256fifo.c \
         $(DRIVERS)/api/no_os_irq.c      \
-         $(DRIVERS)/api/no_os_timer.c   \
+        $(DRIVERS)/api/no_os_timer.c    \
```
nit: unrelated change.

rbudai98 replied:
but it was odd :))

amiclaus commented on projects/adt75/src.mk:
```diff
@@ -23,7 +23,8 @@ INCS += $(INCLUDE)/no_os_delay.h     		\
 		$(INCLUDE)/no_os_units.h        \
-		$(INCLUDE)/no_os_alloc.h
+		$(INCLUDE)/no_os_alloc.h		\
+        $(INCLUDE)/no_os_mutex.h
```
odd indentation.

amiclaus commented on projects/eval-adis/src.mk:
```diff
@@ -24,7 +24,8 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_units.h \
-		$(INCLUDE)/no_os_alloc.h
+		$(INCLUDE)/no_os_alloc.h \
+        $(INCLUDE)/no_os_mutex.h
```
ditto: odd indentation.

amiclaus commented on projects/eval-adxl313z/src.mk:
```diff
@@ -24,7 +24,8 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_init.h \
-		$(INCLUDE)/no_os_alloc.h
+		$(INCLUDE)/no_os_alloc.h \
+        $(INCLUDE)/no_os_mutex.h
```
ditto.

amiclaus commented on projects/eval-adxl313z/src.mk:
```diff
@@ -34,7 +35,8 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(NO-OS)/util/no_os_util.c \
-		$(NO-OS)/util/no_os_alloc.c
+		$(NO-OS)/util/no_os_alloc.c \
+        $(NO-OS)/util/no_os_mutex.c
```
ditto

amiclaus commented on projects/eval-adxl355-pmdz/src.mk:
```diff
@@ -25,7 +25,8 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_init.h \
-		$(INCLUDE)/no_os_alloc.h
+		$(INCLUDE)/no_os_alloc.h \
+        $(INCLUDE)/no_os_mutex.h
```
ditto

amiclaus commented on projects/eval-adxl355-pmdz/src.mk:
```diff
@@ -36,7 +37,8 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(NO-OS)/util/no_os_util.c \
-		$(NO-OS)/util/no_os_alloc.c
+		$(NO-OS)/util/no_os_alloc.c \
+        $(NO-OS)/util/no_os_mutex.c
```
ditto.

amiclaus commented on projects/max31855/src.mk:
```diff
@@ -22,7 +22,8 @@ INCS += $(INCLUDE)/no_os_delay.h     \
 		$(INCLUDE)/no_os_util.h 	\
-		$(INCLUDE)/no_os_units.h 
+		$(INCLUDE)/no_os_units.h	\
+        $(INCLUDE)/no_os_mutex.h 
```
ditto.

amiclaus commented on projects/max31855/src.mk:
```diff
@@ -31,7 +32,8 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 		$(NO-OS)/util/no_os_util.c \
-		$(NO-OS)/util/no_os_alloc.c
+		$(NO-OS)/util/no_os_alloc.c \
+        $(NO-OS)/util/no_os_mutex.c
```
ditto.

amiclaus reviewed (commented):
some comments on my side.

rbudai98 commented:
v3:
* addressed PR comments

rbudai98 commented:
V4:
* removed commit 34ebcab, hence it was added in an other PR
* modified api files according to recent changes in master
* i2c and spi have bus that allocates mutex automatically, while other peripherals have individual mutexes

buha commented on drivers/api/no_os_eeprom.c:
```diff
@@ -41,6 +41,12 @@
+/**
+ * @brief - EEPROM mutex
+*/
+static void *eeprom_mutex;
```
i'm not sure i understand the role of this global

buha replied:
(same comment for all the other files)

rbudai98 replied:
This pointer is initialized only once to point to the allocated mutex. Otherwise is just a null pointer.

rbudai98 replied:
Similar to this, but having only one element:
https://github.com/analogdevicesinc/no-OS/blob/dea741fba32637889047a3eba778565f3e3ecfe3/drivers/api/no_os_i2c.c#L50

rbudai98 replied:
Will change it to array

rbudai98 commented:
v5:
* changed mutex pointers to array of pointers and added header defffinitions for max number of peripherals
* drivers/api/spi: transfer return value corrected
* include/no_os_mutex.h: added warning to header functions

buha commented on drivers/api/no_os_eeprom.c:
```diff
@@ -64,6 +70,9 @@ int32_t no_os_eeprom_init(struct no_os_eeprom_desc **desc,
 	(*desc)->platform_ops = param->platform_ops;
+	(*desc)->mutex = eeprom_mutex_table[param->device_id];
+
+	no_os_mutex_init((*desc)->mutex);
```
Do these inits actually work ?
I think they should be `no_os_mutex_init(&(*desc)->mutex);` (*address of a pointer*)

rbudai98 replied:
added pointer reference

buha commented on include/no_os_eeprom.h:
```diff
@@ -50,6 +50,8 @@
 /********************** Macros and Constants Definitions **********************/
 /******************************************************************************/
 
+#define EEPROM_MAX_NUMBER 4
```
For Pwm/Timer/Uart it's pretty clear, they are hw peripherals that have a "device id" and having a static table with mutexes for each makes sense. But in the case of EEPROM and MDIO, it's not clear to me how to handle them. I would propose to remove the eeprom/mdio changes from this PR, think them through and add them later, separately. For me t's not even clear what an EEPROM device ID is, and that ID is used as a index in the mutex table. So I think we need to sort that out in the team first.

rbudai98 replied:
removed

rbudai98 commented:
v6:
* added dereferencing to the mutex initialization
* removed EEPROM and MDIO mutex implementation
* added slave counter and counter logic to spi and i2c bus descriptors

buha commented on drivers/api/no_os_spi.c:
```diff
@@ -138,11 +140,15 @@ void no_os_spibus_remove(uint32_t bus_number)
 			spi_table[bus_number];
 
-	no_os_mutex_remove(spi_bus_desriptor->mutex);
+	spi_bus_desriptor->slave_number = spi_bus_desriptor->slave_number - 1;
```
you can simplify by doing `spi_bus_desriptor->slave_number--;`

buha replied:
also, since we're here, you should only do `slave_number--` if `slave_number > 0`, otherwise you can end up with a wrap around

buha replied:
these comments are valid both for spi and i2c

buha commented:
One more thing, I think you need to changes this in no_os_spibus_remove():

```
no_os_spibus_remove(desc->device_id);
```

to this:
```
no_os_spibus_remove(desc->bus->device_id);
```

https://github.com/analogdevicesinc/no-OS/blob/23a85b3c04b13b5a67f0e2cff88b40f9745a4b46/drivers/api/no_os_spi.c#L124

Valid also for i2c.

rbudai98 commented:
v7:
* applied PR suggestions

buha commented on drivers/api/no_os_i2c.c:
```diff
@@ -137,11 +139,16 @@ void no_os_i2cbus_remove(uint32_t bus_number)
 			i2c_table[bus_number];
 
-	no_os_mutex_remove(i2c_bus_desriptor->mutex);
+	if (i2c_bus_desriptor->slave_number > 0)
```
rename `desriptor` -> `descriptor` or delete it altogether - it makes for an unnecessarily long name, it could just as well be a variable named `bus`

buha reviewed (approved):
looks good now, i am merging
